Title : Multiple linear regression - Without built-in functions<br>
Author : raghav | Github : @raghavtwenty <br>
Date Created : April 4, 2023 | Last Updated : April 11, 2023 <br>
Python Version : 3.11.2


AIM : <br>
To identify the relevant features that cause the rainfall and predict the intensity of the daily rainfall using regression models(either Linear regression/multivariate regression). Use Rain Fall dataset.


ALGORITHM : <br>
Step 01 : Start the process.<br>
Step 02 : Import the Rain fall dataset.<br>
Step 03 : Check for null values.<br>
Step 04 : Convert the string attributes to int values.<br>
Step 05 : Fit the multiple regression model.<br>
Step 06 : Calculate the measure of accuracy for rainfall prediction using environmental variables.<br>
Step 07 : Stop the process.<br>

<br> CODE & OUTPUT<br>


Imports


In [206]:
import pandas as pd
from sklearn.model_selection import train_test_split
import math


Load the dataset


In [207]:
df = pd.read_csv(
    "/Users/raghav/Documents/Programs/python/ml/datasets/Rainfall_india_1901_2015.csv"
)


In [208]:
print(df.columns)

Index(['SUBDIVISION', 'YEAR', 'JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL',
       'AUG', 'SEP', 'OCT', 'NOV', 'DEC', 'ANNUAL', 'Jan-Feb', 'Mar-May',
       'Jun-Sep', 'Oct-Dec'],
      dtype='object')


In [209]:
df.tail()

,SUBDIVISION,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANNUAL,Jan-Feb,Mar-May,Jun-Sep,Oct-Dec
4111,LAKSHADWEEP,2011,5.1,2.8,3.1,85.9,107.2,153.6,350.2,254.0,255.2,117.4,184.3,14.9,1533.7,7.9,196.2,1013.0,316.6
4112,LAKSHADWEEP,2012,19.2,0.1,1.6,76.8,21.2,327.0,231.5,381.2,179.8,145.9,12.4,8.8,1405.5,19.3,99.6,1119.5,167.1
4113,LAKSHADWEEP,2013,26.2,34.4,37.5,5.3,88.3,426.2,296.4,154.4,180.0,72.8,78.1,26.7,1426.3,60.6,131.1,1057.0,177.6
4114,LAKSHADWEEP,2014,53.2,16.1,4.4,14.9,57.4,244.1,116.1,466.1,132.2,169.2,59.0,62.3,1395.0,69.3,76.7,958.5,290.5
4115,LAKSHADWEEP,2015,2.2,0.5,3.7,87.1,133.1,296.6,257.5,146.4,160.4,165.4,231.0,159.0,1642.9,2.7,223.9,860.9,555.4


In [210]:
df.drop(
    [
        "JAN",
        "FEB",
        "MAR",
        "APR",
        "MAY",
        "JUL",
        "JUN",
        "AUG",
        "SEP",
        "OCT",
        "NOV",
        "DEC",
        "Jan-Feb",
        "Mar-May",
        "Jun-Sep",
        "Oct-Dec",
    ],
    inplace=True,
    axis=1,
)


In [211]:
df.head()

,SUBDIVISION,YEAR,ANNUAL
0,ANDAMAN & NICOBAR ISLANDS,1901,3373.2
1,ANDAMAN & NICOBAR ISLANDS,1902,3520.7
2,ANDAMAN & NICOBAR ISLANDS,1903,2957.4
3,ANDAMAN & NICOBAR ISLANDS,1904,3079.6
4,ANDAMAN & NICOBAR ISLANDS,1905,2566.7


Convert the state names into integers


In [212]:
unique_states = sorted(df["SUBDIVISION"].unique())


In [213]:
for col_val in df["SUBDIVISION"]:
    for state in unique_states:
        repl = unique_states.index(state) + 1
        if state == col_val:
            df["SUBDIVISION"] = df["SUBDIVISION"].replace([col_val], repl)


Check


In [214]:
df.head()

,SUBDIVISION,YEAR,ANNUAL
0,1,1901,3373.2
1,1,1902,3520.7
2,1,1903,2957.4
3,1,1904,3079.6
4,1,1905,2566.7


Drop null values


In [215]:
df.dropna(inplace=True)


Selecting the features


In [216]:
x = df[["SUBDIVISION", "YEAR"]]
y = df["ANNUAL"]


In [217]:
x_train, x_test, y_train, y_test = train_test_split(
    x, y, train_size=0.8, test_size=0.2)
print(y_test)


1385     306.6
1510     403.7
773     1633.8
3717     748.6
262     2802.9
         ...  
1516     805.3
3834     992.8
3197     827.7
2982    1443.0
1758    1520.3
Name: ANNUAL, Length: 818, dtype: float64


In [218]:
x1 = x_train["SUBDIVISION"]
x2 = x_train["YEAR"]
y = y_train


In [219]:
length = len(x1)
print(length)

3272


In [220]:
def mlr(x1, x2, y):
    global b0, b1, b2

    print("\nSum")
    x1_sum = x1.sum()
    print(x1_sum)
    x2_sum = x2.sum()
    print(x2_sum)
    y_sum = y.sum()
    print(y_sum)

    print("\nMean")
    x1_mean = x1.mean()
    print(x1_mean)
    x2_mean = x2.mean()
    print(x2_mean)
    y_mean = y.mean()
    print(y_mean)

    print("\nSquared")
    x1_sqr = sum([i**2 for i in x1])
    print(x1_sqr)
    x2_sqr = sum([i**2 for i in x2])
    print(x2_sqr)
    print("\nDot product")
    x1y_sqr = x1.dot(y)
    print(x1y_sqr)
    x2y_sqr = x2.dot(y)
    print(x2y_sqr)
    x1x2_sqr = x1.dot(x2)
    print(x1x2_sqr)

    print("\nRegression Sums")
    x1_reg_sum = x1_sqr - ((x1_sum**2) / length)
    print(x1_reg_sum)
    x2_reg_sum = x2_sqr - ((x2_sum**2) / length)
    print(x2_reg_sum)
    x1y_reg_sum = x1y_sqr - ((x1_sum * y_sum) / length)
    print(x1y_reg_sum)
    x2y_reg_sum = x2y_sqr - ((x2_sum * y_sum) / length)
    print(x2y_reg_sum)
    x1x2_reg_sum = x1x2_sqr - ((x1_sum * x2_sum) / length)
    print(x1x2_reg_sum)

    b1 = ((x2_reg_sum * x1y_reg_sum) - (x1x2_reg_sum * x2y_reg_sum)) / (
        (x1_reg_sum * x2_reg_sum) - (x1x2_reg_sum**2)
    )

    b2 = ((x1_reg_sum * x2y_reg_sum) - (x1x2_reg_sum * x1y_reg_sum)) / (
        ((x1_reg_sum * x2_reg_sum) - (x1x2_reg_sum**2))
    )

    b0 = y_mean - (b1 * (x1_mean)) - (b2 * (x2_mean))

    print("\nCoefficients")
    print(
        f"""
    b0 (Intercept) : {b0}
    b1 : {b1}
    b2 : {b2}
    """
    )

    print(f"\nEquation : {b0} + {b1} * x1 + {b2} * x2")


mlr(x1, x2, y)



Sum
60971
6407816
4647624.7

Mean
18.63416870415648
1958.3789731051345
1420.423196821516

Squared
1486553
12552528356

Dot product
75001055.3
9099485246.9
119405117

Regression Sums
350409.0999388753
3596238.073348999
-11603567.433404654
-2325240.4640598297
792.6308068484068

Coefficients

    b0 (Intercept) : 3289.402176607705
    b1 : -33.112897822491625
    b2 : -0.6392775211905699
    

Equation : 3289.402176607705 + -33.112897822491625 * x1 + -0.6392775211905699 * x2


Estimating the target value for the input subdivision and year


In [249]:
user_x1 = int(input("Enter subdivision : "))
user_x2 = int(input("Enter year : "))
eqn = b0 + (b1 * user_x1) + (b2 * user_x2)
print(eqn)


1963.0308534166904


In [229]:
for i in y_test:
    with open("true.txt", "a") as w:
        w.write(str(i))
        w.write(",")

print(len(y_test))

818


In [230]:
y_pred_user = []  # difference between trained and predicted


def pred(xt):
    p_x1 = xt["SUBDIVISION"]
    p_x2 = xt["YEAR"]

    for i in range(len(p_x1)):
        pred = b0 + (b1 * p_x1.iloc[i]) + (b2 * p_x2.iloc[i])
        # print(p_y.iloc[i], pred)
        with open("predict.txt", "a") as w:
            w.write(str(pred))
            w.write(",")

        y_pred_user.append(pred)


pred(x_test)
print(len(y_pred_user))


818


Finding the RMSE value


In [243]:
diff_list_numerator = []


def diff_ytest_ypred(y_test, y_pred_user, mean):
    for i in range(len(y_pred_user)):
        diff_list_numer = (y_test.iloc[i] - y_pred_user[i]) ** 2
        diff_list_numerator.append(diff_list_numer)


mean = sum(y_pred_user) / len(y_pred_user)
diff_ytest_ypred(y_test, y_pred_user, mean)
n = len(diff_list_numerator)

print(f"RMSE : {math.sqrt(sum(diff_list_numerator)/n)}")

RMSE : 791.8844718585322


---


Validation with the built in function


In [233]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

model = LinearRegression().fit(x_train, y_train)

y_pred = model.predict(x_test)

or_rmse = r2_score(y_test, y_pred)
print(f"ORG R2 Score : {or_rmse}")

ORG R2 Score : 0.12635457221224167


In [247]:
from sklearn.metrics import mean_squared_error

mse = mean_squared_error(y_test, y_pred_user)
rmse = math.sqrt(mse)
print(f"ORG RMSE : {rmse}")

ORG RMSE : 791.884471858532


In [241]:
# print(y_pred)


In [226]:
print(model.coef_)
print(model.intercept_)


[-33.11289782  -0.63927752]
3289.4021766069054


Inference


The RMSE for the MLR model is 791.88 and R^2 Score is 0.126 <br>
When the location is ANDAMAN & NICOBAR ISLANDS and the year is 2023 <br>
that predicted annual rainfall is 1963.03.